docs/src/app/components/AppBar/ExampleWaterfallOptimized.jsx

```diff
@@ -0,0 +1,93 @@
+import React from 'react';
+import {AppBar} from 'material-ui';
+
+import IconButton from 'icon-button';
+import MoreVertIcon from 'svg-icons/navigation/more-vert';
+import ArrowBack from 'svg-icons/navigation/arrow-back';
+
+const MIN_HEIGHT = 64;
+const MAX_HEIGHT = 210;
+
+const AppBarWaterfallExample = React.createClass({
+
+  propTypes: {
+    onBack: React.PropTypes.func,
+  },
+
+  render() {
+    const styles = this.getStyles();
+    return (
+      <AppBar
+        position="waterfall"
+        waterfall={{
+          minHeight: MIN_HEIGHT,
+          maxHeight: MAX_HEIGHT,
+          onHeightChange: this.onHeightChange,
+          children: (<div
+            style={styles.logoWrap}>
+            <img
+              ref={el => { this.logoEl = el; }}
```

Comment:
Having a function with height as parameter seems a good idea at first. I started to implement this feature exactly this way, but it didn't help the performance too much. I even tried react-motion, but still not better. My use-case was to use it for a web app and I really wanted a smooth > 30fps transition on a mid-range phone. The only way to do this is by directly modifying the styles on DOM elements, unfortunately. Probably this is why LeftNav have a similar approach. I think one serious bottleneck with a seState -> render solution is the current inline styling with aggressive and mostly unnecessary prefixing. Maybe some immutable data structures or memoization could help, but that's a long shot. 
Anyway onHeightChange is still necessary for transitions on elements that are not part of the waterfall children, like AppBar title, or, in my app I had a FAB that really needed this callback.